source/opt/local_ssa_elim_pass.cpp

```diff
@@ -0,0 +1,759 @@
+    ir::Instruction* di = deadInsts.front();
+    // Don't delete labels
+    if (di->opcode() == SpvOpLabel) {
+      deadInsts.pop();
+      continue;
+    }
+    // Remember operands
+    std::vector<uint32_t> ids;
+    di->ForEachInId([&ids](uint32_t* iid) {
+      ids.push_back(*iid);
+    });
+    uint32_t varId = 0;
+    // Remember variable if dead load
+    if (di->opcode() == SpvOpLoad)
+      (void) GetPtr(di, &varId);
+    KillNamesAndDecorates(di);
+    def_use_mgr_->KillInst(di);
+    // For all operands with no remaining uses, add their instruction
+    // to the dead instruction queue.
+    for (auto id : ids)
+      if (HasOnlyNamesAndDecorates(id))
+        deadInsts.push(def_use_mgr_->GetDef(id));
+    // if a load was deleted and it was the variable's
+    // last load, add all its stores to dead queue
+    if (varId != 0 && !IsLiveVar(varId)) 
+      AddStores(varId, &deadInsts);
+    deadInsts.pop();
+  }
+}
+
+bool LocalSSAElimPass::HasOnlySupportedRefs(uint32_t varId) {
+  if (supported_ref_vars_.find(varId) != supported_ref_vars_.end())
+    return true;
+  analysis::UseList* uses = def_use_mgr_->GetUses(varId);
+  assert(uses != nullptr);
+  for (auto u : *uses) {
+    const SpvOp op = u.inst->opcode();
+    if (op != SpvOpStore && op != SpvOpLoad && op != SpvOpName &&
+        !IsDecorate(op))
+      return false;
+  }
+  supported_ref_vars_.insert(varId);
+  return true;
+}
+
+void LocalSSAElimPass::InitSSARewrite(ir::Function& func) {
+  // Init predecessors
+  label2preds_.clear();
+  for (auto& blk : func) {
+    uint32_t blkId = blk.id();
+    blk.ForEachSuccessorLabel([&blkId, this](uint32_t sbid) {
+      label2preds_[sbid].push_back(blkId);
+    });
+  }
+  // Remove variables with non-load/store refs from target variable set
+  for (auto& blk : func) {
+    for (auto& inst : blk) {
+      switch (inst.opcode()) {
+        case SpvOpStore:
+        case SpvOpLoad: {
+          uint32_t varId;
+          (void) GetPtr(&inst, &varId);
+          if (!IsTargetVar(varId))
+            break;
+          if (HasOnlySupportedRefs(varId))
+            break;
+          seen_non_target_vars_.insert(varId);
+          seen_target_vars_.erase(varId);
+        } break;
+        default:
+          break;
+      }
+    }
+  }
+}
+
+uint32_t LocalSSAElimPass::MergeBlockIdIfAny(const ir::BasicBlock& blk) {
+  auto merge_ii = blk.cend();
+  --merge_ii;
+  uint32_t mbid = 0;
+  if (merge_ii != blk.cbegin()) {
+    --merge_ii;
+    if (merge_ii->opcode() == SpvOpLoopMerge)
+      mbid = merge_ii->GetSingleWordInOperand(kLoopMergeMergeBlockIdInIdx);
+    else if (merge_ii->opcode() == SpvOpSelectionMerge)
+      mbid = merge_ii->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+  }
+  return mbid;
+}
+
+void LocalSSAElimPass::ComputeStructuredSuccessors(ir::Function* func) {
+  for (auto& blk : *func) {
+    // If no predecessors in function, make successor to pseudo entry
+    if (label2preds_[blk.id()].size() == 0)
```

Comment:
Well, I felt like I couldn't just ignore these blocks. If the user doesn't want them processed, the user should delete them.